src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -792,77 +803,123 @@ static int parse_imgpath(const std::string &imgpath)
+  assert (ifs.is_open());
+  ifs >> cmdline;
+
+  for (unsigned i = 0; i < cmdline.size(); i++) {
+    const char *arg = &cmdline[i];
+    if (i == 0) {
+      if (strcmp(basename(arg) , "rbd-nbd") != 0) {
+        return -EINVAL;
+      } 
+    } else { 
+      args.push_back(arg);    
+    }
+    
+    while (cmdline[i] != '\0') {
+      i++;
+    }
   }
 
+  std::string err_msg;
+  r = parse_args(args, err_msg);
+  return r;
+}
+
+static int get_map_pid(const std::string& pid_path)
+{
+  int pid = 0;
+  std::ifstream ifs;
+  ifs.open(pid_path.c_str(), std::ifstream::in);
+  if (!ifs.is_open()) {
+    return 0;
+  }
+  ifs >> pid;
+  return pid;
+}
+
+static int do_list_mapped_devices()
+{
+  bool r;
+  bool should_print = false;
+  int index = 0;
+  int pid = 0;
+  TextTable tbl;
+
+  tbl.define_column("pid", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
+
   while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
+    std::string nbd_path = "/sys/block/nbd" + stringify(index);
+    if(access(nbd_path.c_str(), F_OK) != 0) {
       break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
-  }
+    }
+    std::string pid_path = nbd_path + "/pid";
+    pid = get_map_pid(pid_path);
 
-  close(fd[0]);
-  close(fd[1]);
+    if(pid > 0) {
+      r = get_mapped_info(pid);
+      if (r < 0) {
+        index++;
+        continue;
+      } 
+      should_print = true;
+      if (snapname.empty()) {
+        snapname = "-";
+      }
+      tbl << pid << poolname << imgname << snapname 
+          << "/dev/nbd" + stringify(index) << TextTable::endrow;
+    }
+
+    index++;
+  }
 
+  if (should_print) {
+    cout << tbl;
+  }
   return 0;
 }
 
-static int rbd_nbd(int argc, const char *argv[])
+static int parse_args(vector<const char*>& args, std::string& err_msg)
 {
-  int r;
-  enum {
-    None,
-    Connect,
-    Disconnect,
-    List
-  } cmd = None;
-
-  vector<const char*> args;
-
-  argv_to_vec(argc, argv, args);
-  md_config_t().parse_argv(args);
-
   std::vector<const char*>::iterator i;
   std::ostringstream err;
 
+  md_config_t().parse_argv(args);
+
   for (i = args.begin(); i != args.end(); ) {
     if (ceph_argparse_flag(args, i, "-h", "--help", (char*)NULL)) {
-      usage();
-      return 0;
+      return -ENODATA;
     } else if (ceph_argparse_witharg(args, i, &devpath, "--device", (char *)NULL)) {
     } else if (ceph_argparse_witharg(args, i, &nbds_max, err, "--nbds_max", (char *)NULL)) {
       if (!err.str().empty()) {
-        cerr << err.str() << std::endl;
-        return EXIT_FAILURE;
+        err_msg = err.str();
```

Comment:
While you are here, it would be good to add "rbd-nbd: " prefix here and when parsing "--max_part" below. It looks like these are only error messages without this prefix. Alternatively you could remove "rbd-nbd: " prefix from all messages in this function and add it when printing the error message in `rbd_nbd()`.